Enterprise-CMCS/macpro-mdct-mfp
# Page Object Model for Playwright tests

### Description
Adding page objects and fixtures for our existing Playwright tests. Hoping that this can be a template for creating readable e2e tests in the Playwright ecosystem.

This is a _precursor_ to completing the test for filling out the WP, because that PR got huge and I would like for us to carefully review and comment on general approach in a more gradual fashion!

### Related ticket(s)
N/A

---
### How to test
- Run the playwright tests 
    - From the _root_ of the MFP project, you can run `yarn test:e2e` or `yarn test:e2e-ui`
- They should all pass


### Notes

#### Some pros (imo)

- functions that are specific to an area of the report (e.g., filling out specific fields) are contained to that area instead of plopped into one bit test
- breaks up the many pages of a report so it feels easier to follow the user's journey in the test
- shorter LOC in the tests because the area-specific functions are abstracted away into the page objects

#### Some cons (imo)

- it's a lot of extra files, feels like more LOC === more complexity

#### Some questions

- Thing to explore: using [soft assertions](https://playwright.dev/docs/best-practices#use-soft-assertions)
- Should we start thinking about and documenting general e2e best practices and opinions somewhere? 
    - For instance, while researching page object model for this PR, I saw many sources state that a best practice is to not use `expect` in page objects and keep them in the tests. I tried to follow that philosophy here (with the exception of the `isReady()` functions)

#### Resources:
- [Playwright docs: POM](https://playwright.dev/docs/pom)
- [Playwright docs: Fixtures](https://playwright.dev/docs/api/class-fixtures)
- [Github discussion on using the POM](https://github.com/microsoft/playwright/issues/1604) and specifically [this comment](https://github.com/microsoft/playwright/issues/1604#issuecomment-1004711489)
- [Reddit discussion about POM organization](https://www.reddit.com/r/QualityAssurance/comments/1248csz/playwright_framework_best_practicesstructure/)
